### PR TITLE
Initial Cut of Claim [Pension/Burial] Uploads and Processing

### PR DESCRIPTION
**Pull request by markolson**
This adds a lot of functionality to `PersistentAttachment` and `SavedClaim`, turning them into STI models that allow us to add custom data and behavior for each type of data, while retaining any shared functionality. Upon submission of a `SavedClaim`, any `PersistentAttachment` that are referenced in the payload are associated with the Claim and their `Workflow` is started, which will eventually send them to a final internal backend system for claim processing. Some of the Tasks were changed to support working with S3 (or at least a 'test' S3 I ran locally).

This also stubs out the work to generate a PDF from the JSON submission and push it throught the same Workflow as the attachments themselves.

There's cleanup to do on the Controllers - Burials is closer to what I expect the base class to be. Just want to get this into staging before continuing down unnecessary tidying.

### CODE REVIEW

File: db/migrate/20170621025522_create_persistent_attachments.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class CreatePersistentAttachments < ActiveRecord::Migration
+  def change
+    create_table :persistent_attachments do |t|
+      t.uuid :guid
+      t.text :file_data
```
Comment (kreek) [inline db/migrate/20170621025522_create_persistent_attachments.rb]: Does this contain PII that needs encrypting?

File: app/uploaders/claim_documentation/uploader.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  Attacher.validate do
+    validate_virus_free
+    validate_max_size 20.megabytes
+    validate_min_size 1.kilobytes
+    validate_mime_type_inclusion %w(application/pdf) # until conversion is in
+    validate_unlocked_pdf
+  end
+
+  def generate_location(io, context)
+    if io.is_a?(File)
```
Comment (kreek) [inline app/uploaders/claim_documentation/uploader.rb]: nit: could this be a case statement?

File: config/initializers/clamscan.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+ClamScan.configure do |clam_config|
+  clam_config.client_location = `which clamdscan`.strip
```
Comment (kreek) [inline config/initializers/clamscan.rb]: Is this for the difference between local and dev/stage/prod? May be better as a setting.

File: app/models/saved_claim/burial.rb [hunk 4]
```diff
@@ -1,11 +1,29 @@
...
+  ATTACHMENT_KEYS = [:transportationReceipts, :deathCertificate].freeze
+
   def regional_office
     PensionBurial::ProcessingOffice.address_for(open_struct_form.claimantAddress.postalCode)
   end
 
-  def confirmation_number
-    "V-BUR-#{id}#{guid[0..6]}".upcase
+  def to_pdf
+    #     # return FormToPDF::Blahblah.convert(open_struct_form)
```
Comment (kreek) [inline app/models/saved_claim/burial.rb]: Obviously commented out so this may be ignorable :) but would this be extracted into a class? maybe part of the same class/lib that the convert to pdf task uses? Only concern is relying on system calls and reliably catching errors from them.

File: spec/models/saved_claims/pension_spec.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+RSpec.describe SavedClaim::Pension do
+  subject { described_class.new }
+  let(:instance) { FactoryGirl.build(:pension_claim) }
+
+  it_should_behave_like 'saved_claim'
+
+  it 'can generate a pdf' do
+    instance.to_pdf
+    expect(true).to be(true)
```
Comment (kreek) [inline spec/models/saved_claims/pension_spec.rb]: Probably better xit'ed than a pending test that passes

File: spec/models/saved_claims/burial_spec.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+RSpec.describe SavedClaim::Burial do
+  subject { described_class.new }
+  let(:instance) { FactoryGirl.build(:burial_claim) }
+
+  it_should_behave_like 'saved_claim'
+
+  it 'can generate a pdf' do
+    instance.to_pdf
+    expect(true).to be(true)
```
Comment (kreek) [inline spec/models/saved_claims/burial_spec.rb]: same here: probably better xit'ed than a pending test that passes

File: app/models/persistent_attachment.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  belongs_to :saved_claim
+  delegate :original_filename, :size, to: :file
+
+  after_initialize :generate_guid, unless: :guid
+  def generate_guid
+    self.guid = SecureRandom.uuid
+  end
+
+  def process
+    raise 'No Associated Claim!' unless saved_claim
```
Comment (kreek) [inline app/models/persistent_attachment.rb]: This raises a RuntimeError, do we want an explicit error class here?

File: app/models/saved_claim.rb [hunk 8]
```diff
@@ -33,7 +48,12 @@ def form_must_be_string
...
     return unless form_is_string
-
     errors[:form].concat(JSON::Validator.fully_validate(VetsJsonSchema::SCHEMAS[self.class::FORM], parsed_form))
   end
+
+  private
+
+  def attachment_keys
+    []
+  end
```
Comment (saneshark) [inline app/models/saved_claim.rb]: maybe you want? 

```
defined? ATTACHMENT_KEYS ? ATTACHMENT_KEYS : []
```

File: app/uploaders/claim_documentation/uploader.rb (hunk 2 above)
Comment (saneshark) [inline app/uploaders/claim_documentation/uploader.rb]: 👍 it would read better because of the way case makes `#is_a?` unnecessary

Comment (markolson) [inline app/uploaders/claim_documentation/uploader.rb]: Yep

File: db/migrate/20170621025522_create_persistent_attachments.rb (hunk 1 above)
Comment (markolson) [inline db/migrate/20170621025522_create_persistent_attachments.rb]: Hm... Yeah, I guess the filenames themselves could have PII.

File: app/uploaders/file_upload.rb [hunk 9]
```diff
@@ -13,11 +13,16 @@ def initialize(**args)
...
+      @attacher.set(file)
+    else
+      @attacher.assign(file)
+    end
     raise ArgumentError, @attacher.errors.join(',') unless @attacher.errors.blank?
     # Pass in the Shrine-serialized uploaded file to the workflow
     w = workflow.new(@attacher, @options)
-    w.start!(trace: trace)
+    job_id = w.start!(trace: trace)
+    { job_id: job_id, file: @attacher.get }
```
Comment (saneshark) [inline app/uploaders/file_upload.rb]: I'm not terribly familiar with how Shrine works. In the event that something goes wrong, for example ArgumentError, does there need to be a cleanup step?

File: lib/shrine/plugins/validate_virus_free.rb [hunk 10]
```diff
@@ -4,8 +4,14 @@ module Plugins
...
-          cached_path = "#{Rails.root}/#{get.to_io.path}"
-          ClamScan::Client.scan(location: cached_path).safe? || add_error(message) && false
+          if ClamScan.configuration.client_location.blank?
+            Shrine.logger.error('NO VIRUS SCANNING ENABLED')
+            return
+          end
+          cached_path = get.download.path
+          result = ClamScan::Client.scan(location: cached_path)
+          # TODO: Log a the full result to sentry
+          result.safe? || add_error(message) && false
```
Comment (saneshark) [inline lib/shrine/plugins/validate_virus_free.rb]: will the error message trigger one of the Common::Exceptions? Do you need a custom one?

**Comment by markolson** [comment]
@saneshark 
Re: "dynamic mechanism". This PR would really have been better served by being a few separate ones, but I got into the weeds and ended up with this mess. 
* Workflow/Task stuff, which is core functionality not specifically tied to this work, which will eventually drive the EVSS doc uploads as well. The entire idea of that is that it's a dynamic way of storing and transforming files, where each system can use just what it needs. In fact, this Pension&Burial work will switch from SFTP to an API (one call per file) in August.
* STI for Claims. For Claims, we know that there's sort of a base set of functionality. Saving the form so that we can process it - either submitting it asynchronously so we can handle downtime, or converting it to a PDF, or just storing it for support reasons. Then the other things like confirmation numbers, encryption, and whatever else have it make sense to me to make that flexible.
* PersistentUpload STI. Because different Claims have different requirements, we need some glue between the Claims, Files and Workflows.

In short, you can write a Workflow for whatever you want, not necessarily using `VetsShrine` as a subclass if you want to stream a file. I'd even say that using Shrine at all for streaming to remote systems is a bad idea, if you're allowed to stream raw files without stamping them or doing scans.

File: app/controllers/v0/claim_documents_controller.rb [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+    def klass
+      case form_id
+      when '21-527EZ', '21P-530'
+        PersistentAttachment::PensionBurial
+      end
+    end
+
+    def form_id
+      (params[:form_id] || '21p-530').upcase # TODO: Remove default once frontend sends it.
```
Comment (markolson) [inline app/controllers/v0/claim_documents_controller.rb]: This will almost certainly be removed before merging - I think the frontend PR is getting in this morning. Just put it here initially because it was convenient to the problem I was solving that very second.

File: app/models/internal_attachment.rb [hunk 12]
```diff
@@ -13,11 +13,11 @@ def initialize(args = {})
     args.each { |k, v| instance_variable_set("@#{k}", v) }
   end
 
-  def method_missing(method_name)
+  def method_missing(method_name, *arguments, &block)
```
Comment (markolson) [inline app/models/internal_attachment.rb]: Going to leave this for now. It's a helper we can refactor away in a later PR, but this diff is a small bugfix I found in how it was setup and used earlier.

File: app/uploaders/file_upload.rb (hunk 9 above)
Comment (markolson) [inline app/uploaders/file_upload.rb]: It cleans up after itself if there's a validation or other failure. Or should.

File: lib/shrine/plugins/validate_virus_free.rb (hunk 10 above)
Comment (markolson) [inline lib/shrine/plugins/validate_virus_free.rb]: They get wrapped in an `ArgumentError` for now, and I have some work to put them in a custom `Common::Exception` wrapper in progress elsewhere.

File: app/models/saved_claim.rb [hunk 13]
```diff
@@ -6,17 +6,32 @@ class SavedClaim < ActiveRecord::Base
...
+    self.guid ||= SecureRandom.uuid
+    self.form_id = self.class::FORM.upcase
+  end
+
+  def process_attachments!
+    GenerateClaimPDFJob.perform_async(id) if respond_to?(:to_pdf)
+    refs = attachment_keys.map { |key| Array(open_struct_form.send(key)) }.flatten
+    files = PersistentAttachment.where(guid: refs.map(&:confirmationCode))
+    files.update_all(saved_claim_id: id)
+    files.reload.each.map(&:process)
```
Comment (markolson) [inline app/models/saved_claim.rb]: `process` just kicks off a Sidekiq worker per file, so there's no real opportunity to fix everything with a rollback.

File: app/controllers/v0/burial_claims_controller.rb [hunk 14]
```diff
@@ -4,28 +4,36 @@ class BurialClaimsController < ApplicationController
...
 
-    def burial_claim_params
-      params.require(:burial_claim).permit(:form)
+    def short_name
+      'burial_claim'
+    end
+
+    def claim_class
+      SavedClaim::Burial
+    end
```
Comment (aub) [inline app/controllers/v0/burial_claims_controller.rb]: This seems fine, but I'm not sure I see the point in factoring it out like this unless the logic in the controller methods above were going to get pulled out into a module or something. If we're just moving constants around, isn't it a premature optimization?

File: app/models/saved_claim/burial.rb [hunk 15]
```diff
@@ -1,11 +1,14 @@
...
+  PERSISTENT_CLASS = PersistentAttachment::PensionBurial
+  ATTACHMENT_KEYS = [:transportationReceipts, :deathCertificate].freeze
+
   def regional_office
     PensionBurial::ProcessingOffice.address_for(open_struct_form.claimantAddress.postalCode)
   end
 
-  def confirmation_number
-    "V-BUR-#{id}#{guid[0..6]}".upcase
+  def to_pdf
```
Comment (aub) [inline app/models/saved_claim/burial.rb]: Shouldn't it just not respond to to_pdf?

File: app/models/saved_claim/burial.rb [hunk 16]
```diff
@@ -1,11 +1,14 @@
 # frozen_string_literal: true
 class SavedClaim::Burial < SavedClaim
   FORM = '21P-530'
+  CONFIRMATION = 'PEN'
+  PERSISTENT_CLASS = PersistentAttachment::PensionBurial
+  ATTACHMENT_KEYS = [:transportationReceipts, :deathCertificate].freeze
```
Comment (aub) [inline app/models/saved_claim/burial.rb]: IMO these sorts of things are a bit cleaner when defined as instance methods that are not implemented on the base class and then return the appropriate values on the inheriting models, but it's another thing that could go either way so it seems fine. It's just that all of the `self.class::ATTACHMENT_KEYS` stuff it introduces into the code feels a little janky and non-idiomatic.

File: app/models/saved_claim/pension.rb [hunk 17]
```diff
@@ -1,11 +1,15 @@
...
+  PERSISTENT_CLASS = PersistentAttachment::PensionBurial
+
   def regional_office
     PensionBurial::ProcessingOffice.address_for(open_struct_form.veteranAddress.postalCode)
   end
 
-  def confirmation_number
-    "V-PEN-#{id}#{guid[0..6]}".upcase
+  def attachment_keys
+    self.class::ATTACHMENT_KEYS
```
Comment (aub) [inline app/models/saved_claim/pension.rb]: Ok, so now this feels a little like it's passing a jankiness threshold. ;) Why not implement it as an instance method in the first place? Or just use the class variable every time you need to reference it?

File: app/uploaders/file_upload.rb [hunk 18]
```diff
@@ -11,13 +11,18 @@ def initialize(**args)
...
 
-  def start!(file, trace: nil)
+  def start!(file, trace: Thread.current['request_id'])
     # run the shrine upload process.
-    @attacher.assign(file)
+    if file.is_a? Shrine::UploadedFile
+      @attacher.set(file)
+    else
+      @attacher.assign(file)
+    end
```
Comment (aub) [inline app/uploaders/file_upload.rb]: Is there not a way to get these two to have the same interface?

File: config/application.rb [hunk 19]
```diff
@@ -18,6 +18,7 @@
 
 module VetsAPI
   class Application < Rails::Application
+    config.active_record.raise_in_transactional_callbacks = true
```
Comment (aub) [inline config/application.rb]: Can we add a comment about what this does?

File: lib/pdf_fill/filler.rb [hunk 20]
```diff
@@ -6,14 +6,14 @@ module PdfFill
...
-    PDF_FORMS = PdfForms.new('pdftk')
+    PDF_FORMS = PdfForms.new(Settings.binaries.pdftk)
     FORM_CLASSES = {
       '21P-527EZ' => PdfFill::Forms::VA21P527EZ
     }.freeze
 
     def combine_extras(old_file_path, extras_generator)
       if extras_generator.text?
-        file_path = "tmp/pdfs/form_#{Time.zone.now}_final.pdf"
+        file_path = "#{old_file_path.gsub('.pdf', '')}_final.pdf"
```
Comment (aub) [inline lib/pdf_fill/filler.rb]: To @kreek's question about PII above, there could be a problem with it in the filename here right? Does this also introduce a slight chance that filenames could collide if two people use the same name?

File: lib/sftp_writer/remote.rb [hunk 21]
```diff
@@ -29,6 +29,24 @@ def write_path
...
   def write(contents, filename)
-    sftp.upload!(StringIO.new(contents), File.join([write_path, filename].compact))
+    path = File.join([write_path, filename].compact)
+    mkdir_safe(path)
+    sftp.upload!(StringIO.new(contents), path)
+  end
+
+  private
+
+  def mkdir_safe(path)
```
Comment (aub) [inline lib/sftp_writer/remote.rb]: This seems worthy of a comment.

File: lib/workflow/runner.rb [hunk 22]
```diff
@@ -60,7 +60,7 @@ def run_task
       # Build the task class with the user and bookkeeping arguments
       task = current_task[:mod].constantize.new(@options, internal: @internal_options)
       runner = task.method(:run)
-      if runner.arity == 1
+      if runner.arity.abs == 1
```
Comment (aub) [inline lib/workflow/runner.rb]: Can it have an arity of -1?????

Comment (markolson) [inline lib/workflow/runner.rb]: Yep! When you define a method with a named arg it goes to `-1`

File: app/controllers/v0/burial_claims_controller.rb (hunk 14 above)
Comment (markolson) [inline app/controllers/v0/burial_claims_controller.rb]: I was really, really hoping to get this merged in last week and push off a refactor that made a shared base controller for Pension and Burial, but, here we are a week later.

File: lib/pdf_fill/filler.rb (hunk 20 above)
Comment (markolson) [inline lib/pdf_fill/filler.rb]: `old_file_path` is `tmp/pdfs//#{saved_claim.form_id}_#{saved_claim.id}.pdf"`, where saved_claim is an instance of `SavedClaim`. As long as primary key ids are unique, we're fine and would only ever potentially overwrite ourselves.

**Comment by markolson** [comment]
@aub Changes made

**Comment by markolson** [comment]
@saneshark, if you're in today, can you take a final look over this?

File: db/migrate/20170621025522_create_persistent_attachments.rb (hunk 1 above)
Comment (saneshark) [inline db/migrate/20170621025522_create_persistent_attachments.rb]: @markolson are you mitigating this somehow?
